libraries/AC_CustomControl/AC_CustomControl.cpp

```diff
@@ -0,0 +1,188 @@
+    _dt(dt),
+    _ahrs(ahrs),
+    _att_control(att_control),
+    _motors(motors)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+void AC_CustomControl::init(void)
+{
+    switch (CustomControlType(_controller_type))
+    {
+        case CustomControlType::CONT_NONE:
+            break;
+        case CustomControlType::CONT_EMPTY: // This is template backend. Don't initialize it.
+            // This is template backend. Don't initialize it.
+            // _backend = new AC_CustomControl_Empty(*this, _ahrs, _att_control, _motors, _dt);
+            // _backend_var_info[get_type()] = AC_CustomControl_Empty::var_info;
+            break;
+        case CustomControlType::CONT_PID:
+            _backend = new AC_CustomControl_PID(*this, _ahrs, _att_control, _motors, _dt);
+            _backend_var_info[get_type()] = AC_CustomControl_PID::var_info;
+            break;
+        default:
+            return;
+    }
+
+    if (_backend && _backend_var_info[get_type()]) {
+        AP_Param::load_object_from_eeprom(_backend, _backend_var_info[get_type()]);
+    }
+}
+
+// run custom controller if it is activated by RC switch and appropriate type is selected
+void AC_CustomControl::update(void)
+{
+    if (is_safe_to_run()) {
+        Vector3f motor_out_rpy;
+
+        motor_out_rpy = _backend->update();
+
+        motor_set(motor_out_rpy);
+    }
+}
+
+// choose which axis to apply custom controller output
+void AC_CustomControl::motor_set(Vector3f rpy) {
+    if (_custom_controller_mask & (uint8_t)CustomControlOption::ROLL) {
+        _motors->set_roll(rpy.x);
+        _att_control->get_rate_roll_pid().set_integrator(0.0);
+    }
+    if (_custom_controller_mask & (uint8_t)CustomControlOption::PITCH) {
+        _motors->set_pitch(rpy.y);
+        _att_control->get_rate_pitch_pid().set_integrator(0.0);
+    }
+    if (_custom_controller_mask & (uint8_t)CustomControlOption::YAW) {
+        _motors->set_yaw(rpy.z);
+        _att_control->get_rate_yaw_pid().set_integrator(0.0);
+    }
+}
+
+// move main controller's target to current states, reset filters,
+// and move integrator to motor output
+// to allow smooth transition to the primary controller
+void AC_CustomControl::reset_main_att_controller(void)
+{
+    // reset attitude and rate target, if feedforward is enabled
+    if (_att_control->get_bf_feedforward()) {
+        _att_control->relax_attitude_controllers();
+    }
+
+    Vector3f gyro_latest = _ahrs->get_gyro_latest();
+
+    _att_control->get_rate_roll_pid().set_integrator(_att_control->rate_bf_targets().x - gyro_latest.x, _motors->get_roll());
+    _att_control->get_rate_pitch_pid().set_integrator(_att_control->rate_bf_targets().y - gyro_latest.y, _motors->get_pitch());
+    _att_control->get_rate_yaw_pid().set_integrator(_att_control->rate_bf_targets().z - gyro_latest.z, _motors->get_yaw());
```

Comment:
I still think these should reset to zero. This could result in massive windup that takes several seconds to clear.

Reply:
After further thinking and checking flight logs my previous experience, the main controller could stabilize the aircraft without resetting the integrator at all or relaxing the target attitude. We already relaxed the target attitude and attitude rate. Until further argument is made towards this, I agree it is safe to reset to zero.